// react-native-ui-lib/src/style/colors.js
import _map from "lodash/map";
import _toLower from "lodash/toLower";
import _toUpper from "lodash/toUpper";
import _inRange from "lodash/inRange";
import _reverse from "lodash/reverse";
import _memoize from "lodash/memoize";
import _indexOf from "lodash/indexOf";
import _findKey from "lodash/findKey";
import _isUndefined from "lodash/isUndefined";
import _isNil from "lodash/isNil";
import _forEach from "lodash/forEach";
//@ts-ignore
import Color from 'color';
import tinycolor from 'tinycolor2';
import { colorsPalette, themeColors } from "./colorsPalette";
import DesignTokens from "./designTokens";
import DesignTokensDM from "./designTokensDM"; //@ts-ignore

import ColorName from "./colorName";
import Scheme from "./scheme";
export class Colors {
  shouldSupportDarkMode = false;

  constructor() {
    const colors = Object.assign(colorsPalette, themeColors);
    Object.assign(this, colors);
    this.loadSchemes({
      light: DesignTokens,
      dark: DesignTokensDM
    });
    Scheme.addChangeListener(() => {
      Object.assign(this, Scheme.getScheme());
    });
  }
  /**
   * Load custom set of colors
   * arguments:
   * colors - map of keys and colors values e.g {grey10: '#20303C', grey20: '#43515C'}
   */


  loadColors(colors) {
    _forEach(colors, (value, key) => {
      this[key] = value;
    });
  }
  /**
   * Load set of schemes for light/dark mode
   * arguments:
   * schemes - two sets of map of colors e.g {light: {screen: 'white'}, dark: {screen: 'black'}}
   */


  loadSchemes(schemes) {
    Scheme.loadSchemes(schemes);
    Object.assign(this, Scheme.getScheme());
  }
  /**
   * Get app's current color scheme
   */


  getScheme() {
    return Scheme.getSchemeType();
  }
  /**
   * Set color scheme for app
   * arguments:
   * scheme - color scheme e.g light/dark/default
   */


  setScheme(scheme) {
    Scheme.setScheme(scheme);
  }
  /**
   * Support listening to Appearance changes
   * and change the design tokens accordingly
   */


  supportDarkMode() {
    this.shouldSupportDarkMode = true;
  }
  /**
   * Add alpha to hex or rgb color
   * arguments:
   * p1 - hex color / R part of RGB
   * p2 - opacity / G part of RGB
   * p3 - B part of RGB
   * p4 - opacity
   */


  rgba(p1, p2, p3, p4) {
    let hex;
    let opacity;
    let red;
    let green;
    let blue; // Handle design token PlatformColor object

    if (typeof p1 === 'object') {
      p1 = colorStringValue(p1);
    }

    if (arguments.length === 2 && typeof p1 === 'string') {
      hex = p1;
      opacity = p2;
      hex = validateHex(hex);
      red = parseInt(hex.substring(0, 2), 16);
      green = parseInt(hex.substring(2, 4), 16);
      blue = parseInt(hex.substring(4, 6), 16);
    } else if (arguments.length === 4 && typeof p1 === 'number') {
      red = validateRGB(p1);
      green = validateRGB(p2);
      blue = validateRGB(p3);
      opacity = p4;
    } else {
      throw new Error('rgba can work with either 2 or 4 arguments');
    }

    return `rgba(${red}, ${green}, ${blue}, ${opacity})`;
  }

  getBackgroundKeysPattern() {
    return /^(bg-|background-)/;
  }

  isEmpty(color) {
    if (_isNil(color) || color === 'transparent') {
      return true;
    }

    try {
      Color(color);
      return false;
    } catch (error) {
      console.warn('Colors.isEmpty failed:', error);
      return true;
    }
  }

  getColorName(colorValue) {
    const color = colorStringValue(colorValue);
    return ColorName.name(color)[1];
  }

  getColorTint(colorValue, tintKey) {
    if (_isUndefined(tintKey) || isNaN(tintKey) || _isUndefined(colorValue)) {
      // console.error('"Colors.getColorTint" must accept a color and tintKey params');
      return colorValue;
    }

    const color = colorStringValue(colorValue);

    if (color === 'transparent') {
      return color;
    }

    const colorKey = _findKey(this, (_value, key) => this[key] === color);

    if (colorKey) {
      const colorKeys = [1, 5, 10, 20, 30, 40, 50, 60, 70, 80];

      const keyIndex = _indexOf(colorKeys, Number(tintKey));

      const key = this.shouldSupportDarkMode && Scheme.getSchemeType() === 'dark' ? colorKeys[colorKeys.length - 1 - keyIndex] : tintKey;
      const requiredColorKey = `${colorKey.slice(0, -2)}${key}`;
      const requiredColor = this[requiredColorKey];

      if (_isUndefined(requiredColor)) {
        return this.getTintedColorForDynamicHex(color, tintKey);
      }

      return requiredColor;
    }

    return this.getTintedColorForDynamicHex(color, tintKey);
  }

  getTintedColorForDynamicHex(color, tintKey) {
    // Handles dynamic colors (non uilib colors)
    let tintLevel = Math.floor(Number(tintKey) / 10);
    tintLevel = Math.max(1, tintLevel);
    tintLevel = Math.min(8, tintLevel);
    const colorsPalette = this.generateColorPalette(color);
    return colorsPalette[tintLevel - 1];
  }

  generateColorPalette = _memoize(color => {
    const hsl = Color(color).hsl();
    const lightness = Math.round(hsl.color[2]);
    const lightColorsThreshold = this.shouldGenerateDarkerPalette(color) ? 5 : 0;
    const ls = [hsl.color[2]];
    let l = lightness - 10;

    while (l >= 20 - lightColorsThreshold) {
      ls.unshift(l);
      l -= 10;
    }

    l = lightness + 10;

    while (l < 100 - lightColorsThreshold) {
      ls.push(l);
      l += 10;
    }

    const tints = [];

    _forEach(ls, e => {
      const tint = generateColorTint(color, e);
      tints.push(tint);
    });

    const sliced = tints.slice(0, 8);
    const adjusted = adjustSaturation(sliced, color);
    const palette = adjusted || sliced;
    return this.shouldSupportDarkMode && Scheme.getSchemeType() === 'dark' ? _reverse(palette) : palette;
  });

  shouldGenerateDarkerPalette(color) {
    const hsl = Color(color).hsl();
    const hue = hsl.color[0];
    return _inRange(hue, 51, 184);
  }

  isDark(colorValue) {
    const color = colorValue === null ? undefined : colorStringValue(colorValue);
    const lum = tinycolor(color).getLuminance();
    return lum < 0.55;
  }

  isValidHex(string) {
    return /(^#[0-9A-F]{6}$)|(^#[0-9A-F]{3}$)/i.test(string);
  }

  getHexString(color) {
    return tinycolor(color).toHexString();
  }

  getHSL(color) {
    return tinycolor(color).toHsl();
  }

  isTransparent(color) {
    return color && _toUpper(color) === _toUpper('transparent');
  }

  areEqual(colorAValue, colorBValue) {
    const colorA = colorStringValue(colorAValue);
    const colorB = colorStringValue(colorBValue);
    return _toLower(colorA) === _toLower(colorB);
  }

}

function colorStringValue(color) {
  return color.toString();
}

function adjustSaturation(colors, color) {
  let array;
  const lightnessLevel = 80;
  const saturationLevel = 60;
  const hsl = Color(color).hsl();
  const lightness = Math.round(hsl.color[2]);

  if (lightness > lightnessLevel) {
    const saturation = Math.round(hsl.color[1]);

    if (saturation > saturationLevel) {
      array = _map(colors, e => e !== color ? addSaturation(e, saturationLevel) : e);
    }
  }

  return array;
}

function addSaturation(color, saturation) {
  const hsl = Color(color).hsl();
  hsl.color[1] = saturation;
  return hsl.hex();
}

function generateColorTint(color, tintLevel) {
  const hsl = Color(color).hsl();
  hsl.color[2] = tintLevel;
  return hsl.hex();
}

function validateRGB(value) {
  if (isNaN(value) || value > 255 || value < 0) {
    throw new Error(`${value} is invalid rgb code, please use number between 0-255`);
  }

  return value;
}

function validateHex(value) {
  if (!/^#([A-Fa-f0-9]{6}|[A-Fa-f0-9]{3})$/.test(value)) {
    throw new Error(`${value} is invalid hex color`);
  }

  value = value.replace('#', '');

  if (value.length === 3) {
    value = threeDigitHexToSix(value);
  }

  return value;
}

function threeDigitHexToSix(value) {
  return value.replace(/./g, '$&$&');
}

const TypedColors = Colors;
const colorObject = new TypedColors();
colorObject.loadColors(colorsPalette);
export default colorObject;
